src/Agent/NewRelic/Agent/Extensions/Providers/Wrapper/Wcf3/MethodInvokerWrapper.cs

```diff
@@ -145,29 +159,57 @@ public AfterWrappedMethodDelegate BeforeWrappedMethod(InstrumentedMethodCall ins
                     {
-                        retrievedHeaders.Add(new KeyValuePair<string, string>(headerName, httpRequestMessage.Headers[headerName]));
-                    }
+                        var retrievedHeaders = new List<KeyValuePair<string, string>>();
 
-                    transaction.AcceptDistributedTraceHeaders(retrievedHeaders, GetHeaderValue, TransportType.HTTP);
+                        var httpRequestMessage = httpRequestMessageObject as HttpRequestMessageProperty;
+                        if (httpRequestMessage != null)
+                        {
+                            foreach (var headerName in httpRequestMessage.Headers.AllKeys)
+                            {
+                                retrievedHeaders.Add(new KeyValuePair<string, string>(headerName, httpRequestMessage.Headers[headerName]));
+                            }
+
+                            if (retrievedHeaders.Any())
+                            {
+                                haveAcceptedTraceHeaders = true;
+                                transaction.AcceptDistributedTraceHeaders(retrievedHeaders, GetHeaderValue, TransportType.HTTP);
+                            }
+                        }
+                    }
 
-                    IEnumerable<string> GetHeaderValue(List<KeyValuePair<string, string>> headers, string key)
+                    //Supporting NetTCP
+                    if (!haveAcceptedTraceHeaders)
                     {
-                        string value = null;
-
-                        value = retrievedHeaders.
-                            Where(header => header.Key.Equals(key)).
-                            FirstOrDefault().Value;
-
-                        return value == null ? null : new string[] { value };
+                        var messageHeaders = OperationContext.Current?.IncomingMessageHeaders;
+                        if (messageHeaders != null && messageHeaders.MessageVersion != MessageVersion.None)
+                        {
+                            var retrievedHeaders = new List<KeyValuePair<string, string>>();
+                            for (var i = 0; i < messageHeaders.Count; i++)
+                            {
+                                try
+                                {
+                                    retrievedHeaders.Add(new KeyValuePair<string, string>(messageHeaders[i].Name, messageHeaders.GetHeader<string>(i)));
+                                }
+                                catch
+                                {
+                                    //gulp
+                                }
+                            }
+
+                            if (retrievedHeaders.Any())
+                            {
+                                haveAcceptedTraceHeaders = true;
+                                transaction.AcceptDistributedTraceHeaders(retrievedHeaders, GetHeaderValue, TransportType.NetTCP);
```

Comment:
We should also change the transport type from `NetTCP` to `Other`, because of the internal agent spec, and because I think this case will also be triggered by other transport types that use a SOAP based payload.